utils/updateable_value.cc

```diff
@@ -93,30 +96,36 @@ updateable_value_base::updateable_value_base::operator=(std::nullptr_t) {
 
 void
 updateable_value_source_base::for_each_ref(std::function<void (updateable_value_base* ref)> func) {
-    for (auto ref : _refs) {
-        func(ref);
+    for (auto refs_per_shard : _refs) {
+        for (auto ref : refs_per_shard) {
+            func(ref);
+        }
     }
 }
 
 updateable_value_source_base::~updateable_value_source_base() {
-    for (auto ref : _refs) {
-        ref->_source = nullptr;
+    for (auto refs_per_shard : _refs) {
+        for (auto ref : refs_per_shard) {
+            ref->_source = nullptr;
+        }
     }
 }
 
 void
 updateable_value_source_base::add_ref(updateable_value_base* ref) const {
-    _refs.push_back(ref);
+    _refs[ref->_owner].push_back(ref);
 }
 
 void
 updateable_value_source_base::del_ref(updateable_value_base* ref) const {
-    _refs.erase(std::remove(_refs.begin(), _refs.end(), ref), _refs.end());
+    auto& refs_for_shard = _refs[ref->_owner];
+    refs_for_shard.erase(std::remove(refs_for_shard.begin(), refs_for_shard.end(), ref), refs_for_shard.end());
 }
```

Comment:
This causes an smp race if ref->_owner != this_shard_id(), since we can have two threads simultaneously updating the same ref_for_shard vector.

I see two solutions, both unappealing:
1. use futures everywhere
2. use standard locks

I would like to restore the property of cross-shard safety without future baggage, so we can put updateable_values in config structs, but I really dislike to mix standard locking with seastar.

Reply:
Yes, I also thought about standard locks, but I'd rather do futures than that, even if this code is effectively only called in startup/shutdown. The question is, what would it take for 2 threads to try and simultaneously update this ref_for_shard? In the traditional use of `updateable_value`, which is "initialize source on shard0 and then use a single instance of it in other shards, then destroy everything on Scylla shutdown", how can we end up with 2 writers? I'm not stating that we can't, I just haven't figured out such scenario and I'm curious.

Reply:
And, again, perhaps making everything future-based won't be that bad, since we already use seastar threads in most of the initialization code, so it's probably a matter of spraying `get()`s in some places.

Reply:
update: making everything future-based *is* ugly, I tried. Inability to have async constructors/destructors makes the class much less convenient, and that was the whole point of having it.